src/language-mode.coffee

```diff
@@ -255,12 +255,32 @@ class LanguageMode
       precedingRow = bufferRow - 1
       return currentIndentLevel if precedingRow < 0
 
-    precedingLine = @buffer.lineForRow(precedingRow)
     desiredIndentLevel = @editor.indentationForBufferRow(precedingRow)
-    desiredIndentLevel += 1 if increaseIndentRegex.testSync(precedingLine) and not @editor.isBufferRowCommented(precedingRow)
+    unless @editor.isBufferRowCommented(precedingRow)
+      precedingLine = @buffer.lineForRow(precedingRow)
+      desiredIndentLevel += 1 if increaseNextIndentRegex.testSync(precedingLine)
+
+    decreaseNextIndentRegex = @decreaseNextIndentRegexForScopeDescriptor(scopeDescriptor)
+    decreaseIndentRegex = @decreaseIndentRegexForScopeDescriptor(scopeDescriptor)
+    return desiredIndentLevel unless decreaseNextIndentRegex or decreaseIndentRegex
+
+    if decreaseNextIndentRegex?.testSync(precedingLine) or (decreaseIndentRegex?.testSync(tokenizedLine.text) and not @editor.isBufferRowCommented(bufferRow))
+      lastIndentingRow = bufferRow
+      while (lastIndentingRow = @buffer.previousNonBlankRow(lastIndentingRow))?
+        continue if @editor.isBufferRowCommented(lastIndentingRow)
 
-    return desiredIndentLevel unless decreaseIndentRegex = @decreaseIndentRegexForScopeDescriptor(scopeDescriptor)
-    desiredIndentLevel -= 1 if decreaseIndentRegex.testSync(tokenizedLine.text)
+        lastIndentingLine = @buffer.lineForRow(lastIndentingRow)
+        if increaseNextIndentRegex.testSync(lastIndentingLine)
+          desiredIndentLevel = @editor.indentationForBufferRow(lastIndentingRow) + 1
+
+          for row in [lastIndentingRow..bufferRow]
+            continue if @editor.isBufferRowCommented(row)
+
+            rowLine = @buffer.lineForRow(row)
+            desiredIndentLevel -= 1 if decreaseNextIndentRegex?.testSync(rowLine)
+            desiredIndentLevel -= 1 if decreaseIndentRegex?.testSync(rowLine)
+
+          return Math.max(desiredIndentLevel, 0)
```

Comment:
@maxbrunsfeld additionally please note that it will only ever get to the `while` and `for` loops when you insert a newline (press return) and the trailing chars that will then move to the newline, match a decrease or decreaseNext pattern.

So this code path itself will be triggered by every newline (return), but in most cases the loops will never be reached.

Still it should be efficient of course! But I do believe it is. I myself (and some colleagues) are using this for some time now and I can tell any delays or something alike. Don't know if there are benchmark tests for something like this to show the impact?